components/generators/kubernetes/__init__.py

```diff
@@ -0,0 +1,667 @@
+    def new(self):
+        self.kwargs.apiVersion = "apps/v1"
+        self.kwargs.kind = "Deployment"
+        super().new()
+        self.need("component")
+
+    def body(self):
+        default_strategy = {
+            "type": "RollingUpdate",
+            "rollingUpdate": {
+                "maxSurge": "25%",
+                "maxUnavailable": "25%"
+            }
+        }
+        super().body()
+        component = self.kwargs.component
+        self.root.spec.template.metadata.labels = self.root.metadata.labels
+        self.root.spec.selector.matchLabels = self.root.metadata.labels
+        self.root.spec.template.spec.restartPolicy = component.get("restart_policy", "Always")
+        self.root.spec.strategy = component.get("strategy", default_strategy)
+
+
+class StatefulSet(k8s.Base, WorkloadCommon):
+    def new(self):
+        self.kwargs.apiVersion = "apps/v1"
+        self.kwargs.kind = "StatefulSet"
+        super().new()
+        self.need("component")
+
+    def body(self):
+        default_strategy = {}
+        update_strategy = {"rollingUpdate":
+                               {"partition": 0},
+                           "type": "RollingUpdate"}
+
+        super().body()
+        name = self.kwargs.name
+        component = self.kwargs.component
+        self.root.spec.template.metadata.labels = self.root.metadata.labels
+        self.root.spec.selector.matchLabels = self.root.metadata.labels
+        self.root.spec.template.spec.restartPolicy = component.get("restart_policy", "Always")
+        self.root.spec.strategy = component.get("strategy", default_strategy)
+        self.root.spec.updateStrategy = component.get("update_strategy", update_strategy)
+        self.root.spec.serviceName = name
+
+
+class Job(k8s.Base, WorkloadCommon):
+    def new(self):
+        self.kwargs.apiVersion = "batch/v1"
+        self.kwargs.kind = "Job"
+        super().new()
+        self.need("component")
+
+    def body(self):
+        super().body()
+        name = self.kwargs.name
+        component = self.kwargs.component
+        self.root.spec.template.metadata.labels = self.root.metadata.labels
+        self.root.spec.selector.matchLabels = self.root.metadata.labels
+        self.root.spec.template.spec.restartPolicy = component.get("restart_policy", "Never")
+        self.root.spec.backoffLimit = component.get("backoff_limit", 1)
+        self.root.spec.completions = component.get("completions", 1)
+        self.root.spec.parallelism = component.get("parallelism", 1)
+
+
+class Container(BaseObj):
+    def new(self):
+        self.need("name")
+        self.need("container")
+
+    def process_envs(self, container):
+        for name, value in sorted(container.env.items()):
+            if isinstance(value, dict):
+                if "secretKeyRef" in value:
+                    value["secretKeyRef"]["name"] = self.kwargs.name
+                    self.root.env += [{"name": name, "valueFrom": value}]
```

Comment:
I don't think this block is quite right - the user should be able to set the secret name rather than it always being overwritten by the component name, like this:
```
      component_defaults:
        env:
          MY_PASSWORD:
            secretKeyRef:
              key: MY_PASSWORD
              name: my-secrets
```

If the name is not set, I would suggest at least logging a visible warning that a default secret name is being used, or fail outright, since attempting to deploy a container with an empty-named secret env var will fail with `spec.template.spec.containers[0].env[0].valueFrom.secretKeyRef.name: Invalid value: ""`

I would suggest either
```
                if "secretKeyRef" in value:
                    if "name" not in value["secretKeyRef"]:
                        # warning message here
                        value["secretKeyRef"]["name"] = self.kwargs.name

Reply:
Thanks, there is a part of the code missing there. 

If you have only 1 secret/configmap, they are automatically named "component_name". If you have multiple secrets/configmaps, they are called "{component-name}-{secret-name}"

The generator should then discover the name of the secret from the key. this way we make it more automatic. But that part of the logic is missing, thank you for catching it. 

I agree the user should also be able to override it and set the correct name

Reply:
@alanhughes I have implemented what I have in mind. Now the code will search the "key" inside the secrets (for secretKeyRef) or config_maps (for configMapKeyRef) and automatically add the name. 

This means that if you refactor the configs  (i.e. split it into multiple configmaps for instance) the generator will do the right things and adapt the name automatically.

As you suggested, I have also added the option to override this logic with an explicit "name" field.